File: dart_sample/lib/demos/general/two_dimensional_scroll_view_demo_page.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import '../../counter_widgets.dart';

class TwoDimensionalScrollViewDemoPage extends StatefulWidget {
  const TwoDimensionalScrollViewDemoPage({super.key});

  @override
  State<TwoDimensionalScrollViewDemoPage> createState() =>
      _TwoDimensionalScrollViewDemoPageState();
}

class _TwoDimensionalScrollViewDemoPageState
    extends State<TwoDimensionalScrollViewDemoPage> {
  static const int columnCount = 20;
  static const int rowCount = 20;

  final ScrollController _verticalController = ScrollController();
  final ScrollController _horizontalController = ScrollController();
  DiagonalDragBehavior _behavior = DiagonalDragBehavior.none;

  @override
  void dispose() {
    _verticalController.dispose();
    _horizontalController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'TwoDimensionalScrollView',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'One viewport scrolled by two positions at once. The diagonal drag behavior '
          'decides whether a drag locks to one axis or moves both.',
          style: TextStyle(fontSize: 14, color: Colors.grey),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _behaviorButton('Locked', DiagonalDragBehavior.none),
            _behaviorButton('Weighted', DiagonalDragBehavior.weightedEvent),
            _behaviorButton('Continuous', DiagonalDragBehavior.weightedContinuous),
            _behaviorButton('Free', DiagonalDragBehavior.free),
          ],
        ),
        Expanded(
          child: SampleTableView(
            delegate: TwoDimensionalChildBuilderDelegate(
              builder: _buildCell,
              maxXIndex: columnCount - 1,
              maxYIndex: rowCount - 1,
            ),
            diagonalDragBehavior: _behavior,
            verticalDetails:
                ScrollableDetails.vertical(controller: _verticalController),
            horizontalDetails:
                ScrollableDetails.horizontal(controller: _horizontalController),
          ),
        ),
      ],
    );
  }

  Widget _behaviorButton(String label, DiagonalDragBehavior behavior) {
    final bool selected = _behavior == behavior;
    return Expanded(
      child: CounterTapButton(
        label: label,
        onTap: () => setState(() => _behavior = behavior),
        background: selected ? Colors.blueGrey : const Color(0xFFB0BEC5),
        foreground: selected ? Colors.white : Colors.black,
        fontSize: 13,
      ),
    );
  }

  static Widget _buildCell(BuildContext context, ChildVicinity vicinity) {
    final bool shaded = (vicinity.xIndex + vicinity.yIndex) % 2 == 0;
    return Container(
      color: shaded ? const Color(0xFFE1F5FE) : const Color(0xFFFFF8E1),
      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      child: Text(
        'R${vicinity.yIndex}:C${vicinity.xIndex}',
        style: const TextStyle(fontSize: 13, color: Colors.black),
      ),
    );
  }
}

/// A minimal grid built on [TwoDimensionalScrollView].
class SampleTableView extends TwoDimensionalScrollView {
  const SampleTableView({
    super.key,
    required TwoDimensionalChildBuilderDelegate super.delegate,
    super.diagonalDragBehavior,
    super.verticalDetails,
    super.horizontalDetails,
  });

  @override
  Widget buildViewport(
    BuildContext context,
    ViewportOffset verticalOffset,
    ViewportOffset horizontalOffset,
  ) {
    return SampleTableViewport(
      verticalOffset: verticalOffset,
      verticalAxisDirection: verticalDetails.direction,
      horizontalOffset: horizontalOffset,
      horizontalAxisDirection: horizontalDetails.direction,
      delegate: delegate as TwoDimensionalChildBuilderDelegate,
      mainAxis: mainAxis,
      clipBehavior: clipBehavior,
    );
  }
}

class SampleTableViewport extends TwoDimensionalViewport {
  const SampleTableViewport({
    super.key,
    required super.verticalOffset,
    required super.verticalAxisDirection,
    required super.horizontalOffset,
    required super.horizontalAxisDirection,
    required TwoDimensionalChildBuilderDelegate super.delegate,
    required super.mainAxis,
    super.clipBehavior,
  });

  @override
  RenderTwoDimensionalViewport createRenderObject(BuildContext context) {
    return RenderSampleTableViewport(
      horizontalOffset: horizontalOffset,
      horizontalAxisDirection: horizontalAxisDirection,
      verticalOffset: verticalOffset,
      verticalAxisDirection: verticalAxisDirection,
      delegate: delegate as TwoDimensionalChildBuilderDelegate,
      mainAxis: mainAxis,
      childManager: context as TwoDimensionalChildManager,
      clipBehavior: clipBehavior,
    );
  }

  @override
  void updateRenderObject(
    BuildContext context,
    covariant RenderSampleTableViewport renderObject,
  ) {
    renderObject
      ..horizontalOffset = horizontalOffset
      ..horizontalAxisDirection = horizontalAxisDirection
      ..verticalOffset = verticalOffset
      ..verticalAxisDirection = verticalAxisDirection
      ..mainAxis = mainAxis
      ..delegate = delegate
      ..clipBehavior = clipBehavior;
  }
}

class RenderSampleTableViewport extends RenderTwoDimensionalViewport {
  RenderSampleTableViewport({
    required super.horizontalOffset,
    required super.horizontalAxisDirection,
    required super.verticalOffset,
    required super.verticalAxisDirection,
    required TwoDimensionalChildBuilderDelegate super.delegate,
    required super.mainAxis,
    required super.childManager,
    super.clipBehavior,
  });

  static const double cellWidth = 140.0;
  static const double cellHeight = 44.0;

  @override
  void layoutChildSequence() {
    final builderDelegate = delegate as TwoDimensionalChildBuilderDelegate;
    final int maxColumnIndex = builderDelegate.maxXIndex ?? 0;
    final int maxRowIndex = builderDelegate.maxYIndex ?? 0;
    final double horizontalPixels = horizontalOffset.pixels;
    final double verticalPixels = verticalOffset.pixels;

    final int leadingColumn = math.max((horizontalPixels / cellWidth).floor(), 0);
    final int leadingRow = math.max((verticalPixels / cellHeight).floor(), 0);
    final int trailingColumn = math.min(
      ((horizontalPixels + viewportDimension.width) / cellWidth).ceil(),
      maxColumnIndex,
    );
    final int trailingRow = math.min(
      ((verticalPixels + viewportDimension.height) / cellHeight).ceil(),
      maxRowIndex,
    );

    double xLayoutOffset = (leadingColumn * cellWidth) - horizontalPixels;
    for (int column = leadingColumn; column <= trailingColumn; column++) {
      double yLayoutOffset = (leadingRow * cellHeight) - verticalPixels;
      for (int row = leadingRow; row <= trailingRow; row++) {
        final vicinity = ChildVicinity(xIndex: column, yIndex: row);
        final RenderBox? child = buildOrObtainChildFor(vicinity);
        if (child != null) {
          child.layout(
            constraints.tighten(width: cellWidth, height: cellHeight),
            parentUsesSize: true,
          );
          parentDataOf(child).layoutOffset = Offset(xLayoutOffset, yLayoutOffset);
        }
        yLayoutOffset += cellHeight;
      }
      xLayoutOffset += cellWidth;
    }

    verticalOffset.applyContentDimensions(
      0.0,
      math.max(cellHeight * (maxRowIndex + 1) - viewportDimension.height, 0.0),
    );
    horizontalOffset.applyContentDimensions(
      0.0,
      math.max(cellWidth * (maxColumnIndex + 1) - viewportDimension.width, 0.0),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/General/TwoDimensionalScrollViewDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Foundation;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/general/two_dimensional_scroll_view_demo_page.dart
// (exact sample parity)

namespace Plumix;

public sealed class TwoDimensionalScrollViewDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new TwoDimensionalScrollViewDemoPageState();
    }
}

internal sealed class TwoDimensionalScrollViewDemoPageState : State
{
    private const int ColumnCount = 20;
    private const int RowCount = 20;

    private readonly ScrollController _verticalController = new();
    private readonly ScrollController _horizontalController = new();
    private DiagonalDragBehavior _behavior = DiagonalDragBehavior.None;

    public override void Dispose()
    {
        _verticalController.Dispose();
        _horizontalController.Dispose();
        base.Dispose();
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("TwoDimensionalScrollView", fontSize: 20, color: Colors.Black),
                new Text(
                    "One viewport scrolled by two positions at once. The diagonal drag behavior "
                    + "decides whether a drag locks to one axis or moves both.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BehaviorButton("Locked", DiagonalDragBehavior.None),
                        BehaviorButton("Weighted", DiagonalDragBehavior.WeightedEvent),
                        BehaviorButton("Continuous", DiagonalDragBehavior.WeightedContinuous),
                        BehaviorButton("Free", DiagonalDragBehavior.Free),
                    ]),
                new Expanded(
                    child: new SampleTableView(
                        @delegate: new TwoDimensionalChildBuilderDelegate(
                            BuildCell,
                            maxXIndex: ColumnCount - 1,
                            maxYIndex: RowCount - 1),
                        diagonalDragBehavior: _behavior,
                        verticalDetails: ScrollableDetails.Vertical(controller: _verticalController),
                        horizontalDetails: ScrollableDetails.Horizontal(controller: _horizontalController))),
            ]);
    }

    private Widget BehaviorButton(string label, DiagonalDragBehavior behavior)
    {
        return new Expanded(
            child: new CounterTapButton(
                label: label,
                onTap: () => SetState(() => _behavior = behavior),
                background: _behavior == behavior ? Colors.SteelBlue : Color.Parse("#FFB0BEC5"),
                foreground: _behavior == behavior ? Colors.White : Colors.Black,
                fontSize: 13));
    }

    private static Widget BuildCell(BuildContext context, ChildVicinity vicinity)
    {
        bool shaded = (vicinity.XIndex + vicinity.YIndex) % 2 == 0;
        return new Container(
            color: shaded ? Color.Parse("#FFE1F5FE") : Color.Parse("#FFFFF8E1"),
            padding: new Thickness(10, 8),
            child: new Text(
                $"R{vicinity.YIndex}:C{vicinity.XIndex}",
                fontSize: 13,
                color: Colors.Black));
    }
}

/// <summary>A minimal grid built on <see cref="TwoDimensionalScrollView"/>.</summary>
internal sealed class SampleTableView : TwoDimensionalScrollView
{
    public SampleTableView(
        TwoDimensionalChildBuilderDelegate @delegate,
        DiagonalDragBehavior diagonalDragBehavior,
        ScrollableDetails verticalDetails,
        ScrollableDetails horizontalDetails,
        Key? key = null) : base(
            @delegate,
            diagonalDragBehavior: diagonalDragBehavior,
            verticalDetails: verticalDetails,
            horizontalDetails: horizontalDetails,
            key: key)
    {
    }

    public override Widget BuildViewport(
        BuildContext context,
        ViewportOffset verticalOffset,
        ViewportOffset horizontalOffset)
    {
        return new SampleTableViewport(
            verticalOffset: verticalOffset,
            verticalAxisDirection: VerticalDetails.Direction,
            horizontalOffset: horizontalOffset,
            horizontalAxisDirection: HorizontalDetails.Direction,
            @delegate: (TwoDimensionalChildBuilderDelegate)Delegate,
            mainAxis: MainAxis,
            clipBehavior: ClipBehavior);
    }
}

internal sealed class SampleTableViewport : TwoDimensionalViewport
{
    public SampleTableViewport(
        ViewportOffset verticalOffset,
        AxisDirection verticalAxisDirection,
        ViewportOffset horizontalOffset,
        AxisDirection horizontalAxisDirection,
        TwoDimensionalChildBuilderDelegate @delegate,
        Axis mainAxis,
        Clip clipBehavior,
        Key? key = null) : base(
            verticalOffset,
            verticalAxisDirection,
            horizontalOffset,
            horizontalAxisDirection,
            @delegate,
            mainAxis,
            clipBehavior: clipBehavior,
            key: key)
    {
    }

    public override RenderTwoDimensionalViewport CreateRenderObject(BuildContext context)
    {
        return new RenderSampleTableViewport(
            horizontalOffset: HorizontalOffset,
            horizontalAxisDirection: HorizontalAxisDirection,
            verticalOffset: VerticalOffset,
            verticalAxisDirection: VerticalAxisDirection,
            @delegate: (TwoDimensionalChildBuilderDelegate)Delegate,
            mainAxis: MainAxis,
            childManager: (ITwoDimensionalChildManager)context,
            clipBehavior: ClipBehavior);
    }

    public override void UpdateRenderObject(BuildContext context, RenderObject renderObject)
    {
        var viewport = (RenderSampleTableViewport)renderObject;
        viewport.HorizontalOffset = HorizontalOffset;
        viewport.HorizontalAxisDirection = HorizontalAxisDirection;
        viewport.VerticalOffset = VerticalOffset;
        viewport.VerticalAxisDirection = VerticalAxisDirection;
        viewport.MainAxis = MainAxis;
        viewport.Delegate = Delegate;
        viewport.ClipBehavior = ClipBehavior;
    }
}

internal sealed class RenderSampleTableViewport : RenderTwoDimensionalViewport
{
    private const double CellWidth = 140.0;
    private const double CellHeight = 44.0;

    public RenderSampleTableViewport(
        ViewportOffset horizontalOffset,
        AxisDirection horizontalAxisDirection,
        ViewportOffset verticalOffset,
        AxisDirection verticalAxisDirection,
        TwoDimensionalChildBuilderDelegate @delegate,
        Axis mainAxis,
        ITwoDimensionalChildManager childManager,
        Clip clipBehavior) : base(
            horizontalOffset,
            horizontalAxisDirection,
            verticalOffset,
            verticalAxisDirection,
            @delegate,
            mainAxis,
            childManager,
            clipBehavior: clipBehavior)
    {
    }

    protected override void LayoutChildSequence()
    {
        var builderDelegate = (TwoDimensionalChildBuilderDelegate)Delegate;
        int maxColumnIndex = builderDelegate.MaxXIndex ?? 0;
        int maxRowIndex = builderDelegate.MaxYIndex ?? 0;
        double horizontalPixels = HorizontalOffset.Pixels;
        double verticalPixels = VerticalOffset.Pixels;

        int leadingColumn = Math.Max((int)Math.Floor(horizontalPixels / CellWidth), 0);
        int leadingRow = Math.Max((int)Math.Floor(verticalPixels / CellHeight), 0);
        int trailingColumn = Math.Min(
            (int)Math.Ceiling((horizontalPixels + ViewportDimension.Width) / CellWidth),
            maxColumnIndex);
        int trailingRow = Math.Min(
            (int)Math.Ceiling((verticalPixels + ViewportDimension.Height) / CellHeight),
            maxRowIndex);

        double xLayoutOffset = (leadingColumn * CellWidth) - horizontalPixels;
        for (int column = leadingColumn; column <= trailingColumn; column++)
        {
            double yLayoutOffset = (leadingRow * CellHeight) - verticalPixels;
            for (int row = leadingRow; row <= trailingRow; row++)
            {
                var vicinity = new ChildVicinity(xIndex: column, yIndex: row);
                RenderBox? child = BuildOrObtainChildFor(vicinity);
                if (child != null)
                {
                    child.Layout(
                        Constraints.Tighten(width: CellWidth, height: CellHeight),
                        parentUsesSize: true);
                    ParentDataOf(child).LayoutOffset = new Point(xLayoutOffset, yLayoutOffset);
                }

                yLayoutOffset += CellHeight;
            }

            xLayoutOffset += CellWidth;
        }

        VerticalOffset.ApplyContentDimensions(
            0.0,
            Math.Max((CellHeight * (maxRowIndex + 1)) - ViewportDimension.Height, 0.0));
        HorizontalOffset.ApplyContentDimensions(
            0.0,
            Math.Max((CellWidth * (maxColumnIndex + 1)) - ViewportDimension.Width, 0.0));
    }
}
